templates/cluster/main.go

```diff
@@ -0,0 +1,133 @@
+// package main holds the implementation of the knapsack template.
+package main
+
+import (
+	"time"
+
+	"github.com/nextmv-io/sdk/cluster/kmeans"
+	"github.com/nextmv-io/sdk/measure"
+	"github.com/nextmv-io/sdk/run"
+	"github.com/nextmv-io/sdk/store"
+)
+
+type input struct {
+	Points        []measure.Point `json:"points"`
+	Clusters      int             `json:"clusters"`
+	Weights       []int           `json:"weights"`
+	MaximumWeight int             `json:"maximum_weight"`
+	MaximumPoints int             `json:"maximum_points"`
+}
+
+type cluster struct {
+	Index    int             `json:"index"`
+	Centroid measure.Point   `json:"centroid"`
+	Points   []measure.Point `json:"points"`
+	Indices  []int           `json:"indices"`
+}
+type output struct {
+	Clusters          []cluster       `json:"clusters"`
+	Feasible          bool            `json:"feasible"`
+	Unassigned        []measure.Point `json:"unassigned"`
+	UnassignedIndices []int           `json:"unassigned_indices"`
+}
+
+func main() {
+	run.Run(solver)
+}
+
+func solver(input input, opts store.Options) (store.Solver, error) {
+	// We start by creating a kmeans model.
+	model, err := kmeans.NewModel(input.Points, input.Clusters)
+	if err != nil {
+		return nil, err
+	}
+
+	// We set the maximum weight and maximum points for each cluster.
+	for _, c := range model.ClusterModels() {
+		c.SetMaximumPoints(input.MaximumPoints)
+
+		_, err = c.SetMaximumSumValue(input.MaximumWeight, input.Weights)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// We create a solver with the model.
+	solver, err := kmeans.NewSolver(model)
+	if err != nil {
+		return nil, err
+	}
+
+	// We create the solve options we will use and set the time limit
+	// and th measure to use.
+	solveOptions := kmeans.NewSolveOptions().
+		SetMeasure(measure.EuclideanByPoint()).
+		SetMaximumDuration(opts.Limits.Duration)
+
+	// We use a store, and it's corresponding Format to report the solution
+	// Doing this allows us to use the CLI runner in the main function.
+	root := store.New()
+
+	// Add initial solution as nil
+	so := store.NewVar[kmeans.Solution](root, nil)
+
+	i := 0
+	root = root.Generate(func(s store.Store) store.Generator {
+		return store.Lazy(func() bool {
+			// only run one state transition in which we solve the cluster model
+			return i == 0
```

Comment:
The mechanics of store here looks very odd to me. Is there another way to do this?

Reply:
hm yes i agree this is a little funny bc it has to be summoned from within store. Am i correct that this will not need to be run from within store once we get general runners out since we can run any algorithm after that point?

Reply:
I copy pasted this from the MIP knapsack template to make it a consistent experience in wrapping standalone engine in a store flow.